test_util/test_launch.py

```diff
@@ -0,0 +1,310 @@
+import json
+from contextlib import contextmanager
+
+import pytest
+import yaml
+from docopt import DocoptExit
+
+import ssh.tunnel
+import test_util
+from pkgpanda.util import load_json
+from test_util.helpers import Host
+from test_util.launch import get_launcher, LauncherError, main
+
+MOCK_SSH_KEY_DATA = 'ssh_key_data'
+MOCK_KEY_NAME = 'my_key_name'
+
+
+def check_cli(cmd):
+    assert main(cmd.split(' ')) == 0, 'Command failed! {}'.format(cmd)
+
+
+def check_success(capsys, tmpdir, config_str):
+    """
+    Runs through the required functions of a launcher and then
+    runs through the default usage of the script for a
+    given config path and info path, ensuring each step passes
+    if all steps finished successfully, this parses and returns the generated
+    info JSON and stdout description JSON for more specific checks
+    """
+    # Test launcher directly first
+    config = yaml.load(config_str)
+    launcher = get_launcher(config['type'], config['provider_info'])
+    info = launcher.create(config)
+    launcher.wait(info)
+    launcher.describe(info)
+    launcher.test(info, 'py.test')
+    launcher.delete(info)
+
+    # add config to disk and make info path for CLI testing
+    config_path = tmpdir.join('my_specific_config.yaml')  # test non-default name
+    config_path.write(config_str)
+    config_path = str(config_path)
+    info_path = str(tmpdir.join('my_specific_info.json'))  # test non-default name
+
+    # Now check launcher via CLI
+    check_cli('create --config-path={} --info-path={}'.format(config_path, info_path))
```

Comment:
please use arg arrays rather than relying on parsing (if for some reason someone had a space in their TMPDIR (happens sometimes on some platforms) then this would break. `check_cli(['create', '--config-path={}'.format(config_path), ['--info-path={}'.format(info_path)]` where check_cli then just passes the arg array to main is about the same difficulty to write and introduces no ambiguity.

The whole codebase to date uses direct argument passing. There are many articles on why you should do it in python around subprocess.